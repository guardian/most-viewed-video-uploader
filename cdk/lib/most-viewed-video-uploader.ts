import type {GuStackProps} from "@guardian/cdk/lib/constructs/core";
import {GuStack} from "@guardian/cdk/lib/constructs/core";
import type {App} from "aws-cdk-lib";
import {aws_ssm, Duration} from "aws-cdk-lib";
import {GuScheduledLambda} from "@guardian/cdk";
import {Architecture, Runtime} from "aws-cdk-lib/aws-lambda";
import {Schedule} from "aws-cdk-lib/aws-events";
import {Effect, PolicyStatement} from "aws-cdk-lib/aws-iam";

export class MostViewedVideoUploader extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const urgentAlarmTopicArn = aws_ssm.StringParameter.fromStringParameterName(this, "urgent-alarm-arn", "/account/content-api-common/alarms/urgent-alarm-topic");
    const nonUrgentAlarmTopicArn = aws_ssm.StringParameter.fromStringParameterName(this, "non-urgent-alarm-arn", "/account/content-api-common/alarms/non-urgent-alarm-topic");

    new GuScheduledLambda(this, "MVV", {
      app: "most-viewed-video-uploader",
      architecture: Architecture.ARM_64,
      description: "Most-viewed video uploader",
      fileName: "most-viewed-video-uploader.jar",
      functionName: `most-viewed-video-uploader-v2-${this.stage}`,
      handler: "com.gu.contentapi.Lambda::handleRequest",
      memorySize: 512,
      monitoringConfiguration: {
        snsTopicName: urgentAlarmTopicArn.stringValue,
        toleratedErrorPercentage: 5,
        numberOfMinutesAboveThresholdBeforeAlarm: 1,
      },
      initialPolicy: [
         new PolicyStatement({
           effect: Effect.ALLOW,
           actions: ["s3:GetObject"],
           resources: [`arn:aws:s3:::content-api-config/most-viewed-video-uploader/${this.stage}/*`],
         }),
         new PolicyStatement({
           effect: Effect.ALLOW,
           actions: ["kinesis:PutRecords"],
           resources: [`arn:aws:kinesis:${this.region}:${this.account}:stream/content-api-most-viewed-video-${this.stage}`]
         }),
      ],
      rules: [
        {
          schedule: Schedule.rate(Duration.hours(1)),
        }
      ],
      runtime: Runtime.JAVA_11,
      timeout: Duration.minutes(1)
    })
  }
}
