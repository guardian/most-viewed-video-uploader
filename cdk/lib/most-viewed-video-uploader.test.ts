import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { MostViewedVideoUploader } from "./most-viewed-video-uploader";

describe("The MostViewedVideoUploader stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new MostViewedVideoUploader(app, "MostViewedVideoUploader", { stack: "content-api", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
