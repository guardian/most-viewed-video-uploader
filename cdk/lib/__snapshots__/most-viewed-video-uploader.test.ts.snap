// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MostViewedVideoUploader stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "nonurgentalarmarnParameter": {
      "Default": "/account/content-api-common/alarms/non-urgent-alarm-topic",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "urgentalarmarnParameter": {
      "Default": "/account/content-api-common/alarms/urgent-alarm-topic",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MVV9695AFDA": {
      "DependsOn": [
        "MVVServiceRoleDefaultPolicy2904B8CB",
        "MVVServiceRole73A60CB7",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "content-api/TEST/most-viewed-video-uploader/most-viewed-video-uploader.jar",
        },
        "Description": "Most-viewed video uploader",
        "Environment": {
          "Variables": {
            "APP": "most-viewed-video-uploader",
            "STACK": "content-api",
            "STAGE": "TEST",
          },
        },
        "FunctionName": "most-viewed-video-uploader-TEST",
        "Handler": "com.gu.contentapi.Lambda::handleRequest",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "MVVServiceRole73A60CB7",
            "Arn",
          ],
        },
        "Runtime": "java11",
        "Tags": [
          {
            "Key": "App",
            "Value": "most-viewed-video-uploader",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/most-viewed-video-uploader",
          },
          {
            "Key": "Stack",
            "Value": "content-api",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "MVVErrorPercentageAlarmForLambda4A9B9EE9": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:sns:",
                {
                  "Ref": "AWS::Region",
                },
                ":",
                {
                  "Ref": "AWS::AccountId",
                },
                ":",
                {
                  "Ref": "urgentalarmarnParameter",
                },
              ],
            ],
          },
        ],
        "AlarmDescription": {
          "Fn::Join": [
            "",
            [
              {
                "Ref": "MVV9695AFDA",
              },
              " exceeded 5% error rate",
            ],
          ],
        },
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "High error % from ",
              {
                "Ref": "MVV9695AFDA",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": [
          {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "Error % of ",
                  {
                    "Ref": "MVV9695AFDA",
                  },
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "MVV9695AFDA",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "FunctionName",
                    "Value": {
                      "Ref": "MVV9695AFDA",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "MVVMVVrate1hour0A958706C": {
      "Properties": {
        "ScheduleExpression": "rate(1 hour)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "MVV9695AFDA",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "MVVMVVrate1hour0AllowEventRuleMostViewedVideoUploaderMVV2682A90A8696B223": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MVV9695AFDA",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "MVVMVVrate1hour0A958706C",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "MVVServiceRole73A60CB7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "most-viewed-video-uploader",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/most-viewed-video-uploader",
          },
          {
            "Key": "Stack",
            "Value": "content-api",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MVVServiceRoleDefaultPolicy2904B8CB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/content-api/TEST/most-viewed-video-uploader/most-viewed-video-uploader.jar",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/content-api/most-viewed-video-uploader",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/content-api/most-viewed-video-uploader/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "MVVServiceRoleDefaultPolicy2904B8CB",
        "Roles": [
          {
            "Ref": "MVVServiceRole73A60CB7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
